pkg/live/inventoryrg.go

```diff
@@ -249,78 +308,432 @@ func (icm *InventoryResourceGroup) GetObject() (*unstructured.Unstructured, erro
 
-	// Get cluster object, if exsists.
-	clusterObj, err := namespacedClient.Get(context.TODO(), invInfo.GetName(), metav1.GetOptions{})
-	if err != nil && !apierrors.IsNotFound(err) {
+		// Need to update ResourceGroup to add labels.
+		addShardedLabel(lastRG, icm.Name())
+
+		_, err = namespacedClient.Update(context.TODO(), lastRG, metav1.UpdateOptions{})
 		return err
 	}
 
-	var appliedObj *unstructured.Unstructured
+	// Simply append managed resources to the cluster's last ResourceGroup object in the
+	// sharded series as it should be able to store the additional resources.
+	if ResourceCount == DisableSharding || getObjCount(lastRG) < ResourceCount {
+		items, _, _ := unstructured.NestedSlice(lastRG.Object, "spec", "resources")
+		for obj := range managedResources {
+			items = append(items, idToUnstructuedResources(obj))
+			delete(managedResources, obj)
 
-	if clusterObj == nil {
-		// Create cluster inventory object, if it does not exist on cluster.
-		appliedObj, err = namespacedClient.Create(context.TODO(), invInfo, metav1.CreateOptions{})
-	} else {
-		// Update the cluster inventory object instead.
-		appliedObj, err = namespacedClient.Update(context.TODO(), invInfo, metav1.UpdateOptions{})
+			// Stop appending to item list if we need to shard to a new ResourceGroup.
+			if len(items) == ResourceCount {
+				break
+			}
+		}
+
+		err = unstructured.SetNestedSlice(lastRG.UnstructuredContent(),
+			items, "spec", "resources")
+		if err != nil {
+			return err
+		}
+		_, err = namespacedClient.Update(context.TODO(), lastRG, metav1.UpdateOptions{})
+		if err != nil {
+			return err
+		}
+	}
+
+	// Create newer ResourceGroups to keep storing more resources when sharding is enabled, and
+	// previous ResourceGroup is full.
+	for len(managedResources) > 0 {
+		newRG := icm.nextOf(lastRG)
+		items := make([]interface{}, 0, min(ResourceCount, len(managedResources)))
+		for obj := range managedResources {
+			items = append(items, idToUnstructuedResources(obj))
+			delete(managedResources, obj)
+
+			if len(items) == ResourceCount {
+				break
+			}
+		}
+		err = unstructured.SetNestedSlice(newRG.UnstructuredContent(),
+			items, "spec", "resources")
+		if err != nil {
+			return err
+		}
+		_, err = namespacedClient.Create(context.TODO(), newRG, metav1.CreateOptions{})
+		if err != nil {
+			return err
+		}
+
+		lastRG = newRG
+	}
+
+	return nil
+}
+
+// createShardedName generates the formatted name in a series of sharded ResourceGroups.
+func createShardedName(name string, idx int) string {
+	return fmt.Sprintf(shardedRGNameFmt, name, len(name), idx)
+}
+
+// nextOf returns a new unstructured object with metadata that attributes it to the
+// next sharded ResourceGroup object.
+func (icm *InventoryResourceGroup) nextOf(curr *unstructured.Unstructured) *unstructured.Unstructured {
+	// Copy over metadata from the current ResourceGroup to use as the base for the new ResourceGroup.
+	newRG := curr.DeepCopy()
+	newRG.SetResourceVersion("")
```

Comment:
Does setting it to empty string mean the same thing as removing it from the map?

Maybe not... I know it's different for server-side apply (Apply type Patch), but maybe not Create.